src/cache.rs

```diff
@@ -0,0 +1,505 @@
+        let input_entry_tasks = inputs.0.iter().cloned().enumerate().map(|(i, input)| {
+            let expected_input = compare_with.map(|manifest| &manifest.inputs.0[i]);
+            async move {
+                let digest = if let Some(input_path) = input.input_path() {
+                    Some(D::get_digest(input_path).await?)
+                } else {
+                    None
+                };
+                let input = InputEntry {
+                    key: input.clone(),
+                    value: digest,
+                };
+
+                if let Some(expected_input) = expected_input {
+                    if *expected_input != input {
+                        CacheError::miss(format!(
+                            "Differing build inputs.\nSaw {:#?}\nExpected {:#?})",
+                            input, expected_input
+                        ));
+                    }
+                };
+
+                Ok::<_, CacheError>(input)
+            }
+        });
+
+        let inputs = InputMap(futures::future::try_join_all(input_entry_tasks).await?);
+
+        Ok(Self {
+            inputs,
+            output_path,
+            phantom: PhantomData,
+        })
+    }
+
+    // Writes a manifest file to a particular location.
+    async fn write_to(&self, path: &Utf8PathBuf) -> anyhow::Result<()> {
+        let Some(extension) = path.extension() else {
+            bail!("Missing extension?");
+        };
+        if extension != "json" {
+            bail!("JSON encoding is all we know. Write to a '.json' file?");
+        }
+        let serialized = serde_json::to_string(&self)
+            .with_context(|| "Failed to serialize ArtifactManifest to JSON")?;
+
+        let mut f = File::create(path).await?;
+        f.write_all(serialized.as_bytes()).await?;
+        Ok(())
+    }
+
+    // Reads a manifest file to a particular location.
+    //
+    // Does not validate whether or not any corresponding artifacts exist.
+    async fn read_from(path: &Utf8PathBuf) -> Result<Self, CacheError> {
+        let Some(extension) = path.extension() else {
+            return Err(anyhow!("Missing extension?").into());
+        };
+        if extension != "json" {
+            return Err(anyhow!("JSON encoding is all we know. Read from a '.json' file?").into());
+        }
+
+        let mut f = match File::open(path).await {
+            Ok(f) => f,
+            Err(e) => {
+                if matches!(e.kind(), std::io::ErrorKind::NotFound) {
+                    return Err(CacheError::miss(format!("File {} not found", path)));
+                } else {
+                    return Err(anyhow!(e).into());
+                }
+            }
+        };
+        let mut buffer = String::new();
+        f.read_to_string(&mut buffer)
+            .await
+            .map_err(|e| anyhow!(e))?;
+
+        // In the case that we cannot read the manifest, treat it as "missing".
+        // This will force a rebuild anyway.
```

Comment:
I really like this strategy.

Reply:
Thanks! I had a preference for: "In case of error, treat it like a cache miss, and rebuild anyway", because I figured most users don't care *why* their cache is invalid, they just want a package that works.